### ISSUE TITLE
Transformer

### ISSUE THREAD

**Issue by pillarliang** [pillarliang/blog#12]
## Transformer

`Transformer` 是一个序列到序列（Sequence-to-Sequence，Seq2Seq）的模型，输入与输出序列长度之间的关系有两种情况。

- 输入跟输出的长度一样；
- 机器决定输出的长度。

Transformer 继承了 Seq2Seq 作为 Encoder-Decoder结构的框架，但通过自注意力机制实现了网络结构的重大升级。编码器负责将输入序列转换为高维表示，解码器负责根据编码器的输出生成目标序列。

### 应用

- 语音识别、机器翻译与语音翻译
- 语音合成：输入文字、输出声音信号
- 聊天机器人
- 问答任务
- 句法分析
- 多标签分类：同一个东西可以属于多个类

### 优势

**强大的并行计算能力**

1. 自注意力机制(Self-Attention)的计算可以高度并行化。在计算自注意力时,查询、键和值的矩阵乘法可以完全并行计算。
2. Encoder和Decoder都由多个相同的层堆叠而成，这些相同结构的层可以完全并行运行。
3. 训练过程中,一个batch中的各个样本也可以并行处理。不同样本的Encoder和Decoder的计算是相互独立的,不需要顺序操作。
4. 推理过程中，不同的样本也可以同时并行生成输出。这极大地提高了Transformer模型的吞吐量。

**特征提取能力强**

1. 多头自注意力机制中不同的头可以关注输入的不同部分,提取不同的语义特征。

# Transformer 架构

一般的 `seq2seq` 模型会分成编码器（Encoder）和解码器（Decoder），`Encoder` 负责处理输入的序列，再把处理好的结果“丢”给`Decoder`，由`Decoder`决定要输出的序列。
<img width="681" alt="image" src="https://github.com/pillarliang/blog/assets/28833267/8b418956-877d-4365-882c-daa5beb3dee6">

6个 Encoder 部分是串联的（但 encoder 内部不是），输出给 decoder 是并行的。

## Transformer Encoder

Transformer 的 `Encoder` 使用的是 `self-attention`，输入一排向量，输出另外一个同样长度的一排向量。

**作用**：用于对输入序列进行特征提取和表示。

总体：首先，在输入的地方需要加上位置编码（positional encoding）。经过多头注意力（multi-head attention）后，加上残差连接（residual connection）和层归一化。接下来还要经过全连接的前馈神经网络，接着再做一次残差连接和层归一化，这才是一个块的输出，这个块会重复 N 次。

<img width="644" alt="image" src="https://github.com/pillarliang/blog/assets/28833267/75407a3c-df9e-4936-a021-adb80ef452df">

细说：Encoder 里面会分成很多的 `block`，`block` 并不是神经网络的一层。Transformer 里面加入了**残差连接（`residual connection`）**的设计，最左边的向量 $\bold a$ 输入到自注意力层后得到向量  $\bold b$， 输出向量  $\bold b$ 加上其输入向量 $\bold a$ 得到新的输出。 得到残差的结果以后，再做**层归一化（`layer normalization`）**。层归一化会计算输入向量的均值（mean）和 标准差（standard deviation）。得到层归一化的输出以后，该输出才是全连接网络的输入。输入到全连接网络，再做一个残差连接，把全连接网络的输入跟它的输出加起来得到新的输出。接着把残差的结果再做一次层归一化得到的输出才是 `Transformer Encoder` 里面一个 `block` 的输出。

<aside>
✏️ Transformer Encoder 仅用单向上下文。虽然attention 是支持双向的，比如Bert 就利用的这种双向上下文下。

</aside>

<img width="424" alt="image" src="https://github.com/pillarliang/blog/assets/28833267/a7ad1629-8269-4445-93e5-aef9a4dd4d90">

> Transformer 的 `Encoder` 不一定要这样设计，论文 *[[“On Layer Normalization in the Transformer Architecture”*](https://arxiv.org/abs/2003.07845)](https://arxiv.org/abs/2003.07845) 提出了另一种设计，结果比原始的 Transformer 要好。解释了在 Transformers 里面批量归一化不如层归一化的原因，并提出能量归一化（power normalization）。能量归一化跟层归一化性能差不多，甚至好一点。
> 
> 
> 原始的 Transformer 的架构并不是一个最优的设计，永远可以思考看看有没有更好的设计方式。
> 

<aside>
✏️ 可以把编码器看作是一个文本理解器，它通过自注意力和前向传播神经网络运算处理输入序列，让后续模型能够更好地利用输入信息解决实际问题。

如果将输入的序列看作是一段文本，那么编码器的输出就相当于对这段文本进行了处理、提炼，得到了一段更加精炼的向量表示。这个向量表示保留了输入序列中每个位置的上下文信息，相当于是对输入序列的一种“摘要”或“总结”，可以供后续模型进行下一步的任务处理。

</aside>

## Transformer Decoder

提取输入特征 ，获得输入序列表示。

只能看到上文信息, 未来的信息都被mask掉了。
<img width="561" alt="image" src="https://github.com/pillarliang/blog/assets/28833267/3051eecd-efbf-46d6-ae1b-a07974d79bd4">

### autoregressive｜自回归解码器

> 以语音识别为例，输入一段声音，输出一串文字。把一段声音(*“机器学习”*)给编码器，输出会变成**一排向量**。解码器把编码器的输出先“读”进去，接下来解码器产生语音识别的结果
> 

解码器（Decoder）比较常见的叫做**自回归的（`autoregressive`）解码器**。

在模型实际使用时，解码器的输入会逐步构建。开始时只有开始符号，然后随着新词的生成，每个新词都会添加到输出序列中，并作为新的输入送回解码器，直到生成结束符号（如EOS），表示序列结束。这个过程是自回归的，即每一步的输出都依赖于前一步的输出。

对于每步解码器的输入，都会经过mask Multi-Head Attention (以及Add&LN)后与编码器的输入做注意力。

**详细过程：**

解码器的输入是编码器前面步骤的输出。~~它在前一个时间点的输出~~，其会把自己的输出当做接下来的输入。

`Decoder` 把 `Encoder` 的输出先“读”进去。要让`Decoder`产生输出，首先要先给它一个代表开始的特殊符号 <BOS>（`Begin Of Sequence`），接下来`Decoder`会先进行一个 softmax 操作，“吐” 出一个向量，该向量的长度跟词表的大小是一样的。这个向量会给每一个中文字一个分，分数最高的中文字就是最终的输出。该向量里面的值全部加起来，总和是 1。

> `<BOS>`一个特殊的词元（token）。在机器学习里面，假设要处理自然语言处理的问题，每一个词元都可以用一个独热的向量来表示。
> 

<img width="737" alt="image" src="https://github.com/pillarliang/blog/assets/28833267/d236f909-7f09-461f-88c2-2bc96332bbab">

接下来把“机”当成`Decoder`新的输入。根据两个输入：特殊符号 <BOS> 和“机”，`Decoder`输出一个蓝色的向量。蓝色的向量里面会给出每一个中文字的分数，假设“器” 的分数最高，“器”就是输出。`Decoder`接下来会拿“器”当作输入，其看到了 <BOS>、“机”、“器”， 可能就输出“学”。这个过程就反复地持续下去。

> 机器产生完“习” 以后，还可以继续重复一模一样的过程，把“习”当做输入，解码器可能就会输出“惯”，接 下来就一直持续下去，永远都不会停下来。
> 

要让解码器停止运作，需要特别准备一个特别的符号 `<EOS>`。解码器看到编码器输出的嵌入、`<BOS>`、“机”、“器”、“学”、“习”以后，其产生出来的向量里面 `<EOS>` 的概率必须是最大的，于是输出 <EOS>，整个解码器产生序列的过程就结束了。

**Error propagation | 误差传播**

解码器的输入是它在**前一个时间点**的输出，其会把自己的输出当做接下来的输入。

如果解码器有语音识别的错误，它把机器的“器”识别错成天气的“气”，接下来解码器会根据错误的识别结 果产生它想要产生的期待是正确的输出，这会造成**误差传播（`error propagation`）**的问题，一 步错导致步步错，接下来可能无法再产生正确的词汇。
<img width="546" alt="image" src="https://github.com/pillarliang/blog/assets/28833267/a6b6aceb-be81-4cde-884b-ad2cb4fa4129">

在实际应用中输入跟输出长度的关系是非常复杂的，我们无法从输入序列的长度知道输出序列的长度，因此*解码器必须决定输出的序列的长度*。

### Embedding | 词嵌入

[Feature Engineering | 特征工程](https://www.notion.so/Feature-Engineering-2c1675aec139441e9a13466f481c88ea?pvs=21)

### **Positional Encoding | 位置编码**

[Positional Encoding](https://www.notion.so/Positional-Encoding-f55cabf66be844198b5acf83a4a7a088?pvs=21)

### **Residual connection | 残差连接**

输出与输入再做一次加运算。

为了将信息无损耗的传递的更深，来增强模型的拟合能力。

即使某一层的自注意力没有输出有用的新信息，通过残差连接，确保信息仍能向前传递，不会因为单个层的表现不佳而丢失。

### **LayerNorm | 层归一化**

保持稳定性，并避免梯度消失或爆炸的问题。

随着网络层数的增加，通过多层的计算后的参数可能开始出现过大或过小的情况，模型可能收敛非常的慢，所以进行规范化，使其特征数值在合理范围内。

对单个样本的所有特征进行归一化。

> **batch normalization 不适合处理序列长度可变的数据，而 Transformer 模型通常处理这种数据。**
> 
- **Post-LN**: LN应用在自注意力（Self-Attention）和前馈（Feed Forward）层之后，紧接着残差连接。这在原始的Transformer模型和GPT-1中使用。
- **Pre-LN**: LN应用在自注意力和前馈层之前，之后紧接着残差连接。这在GPT-2和GPT-3等后续模型中更为常见

### **Multi-head attention**

让每个 `attention` 去优化每个词汇的不同特征部分，从而均衡同一种注意力机制可能产生的偏差，让词义拥有来自更多元的表达。

- Encoder 中的是 **self-attention**。
- Decoder 中的是 **cross-attention**。
    
    查询(Q)来自Encoder的输出，而键(K)和值(V)来自另一个序列 Decoder 的输出。
    

### **Masked self-attention**

解码器使用了**掩蔽自注意力（`masked self-attention`）**，掩蔽自注意力可以通过一个**掩码（mask）**来阻止每个位置选择其后面的输入信息。

**`self-attention`**每个向量都要看过完整的输入以后才做决定。**`masked self-attention`**的不同点是不能再看右边的部分，比如：产生 $\bold b^1$ 的时候，只能考虑 $\bold a^1$ 的信息，不能再考虑  $\bold a^2$、$\bold a^3$、 $\bold a^4$。

> Q: 为什么需要在注意力中加掩码?
> 
> 
> A: 解码器的输出是一个一个产生的，所以只能考虑其左边的东西， 没有办法考虑其右边的东西。
> 

[Mask | 掩码](https://www.notion.so/Mask-dc1143d6ba204067a089447c1de139ac?pvs=21)
### **Feed forward | 前馈全链接层**

由两个线性变换和激活函数(通常是ReLU)组成。非线性激活函数位于两层线性变换之间。

引入非线性因素，增强模型的表达能力。

> 这是因为注意力层和其他线性变换本身都是线性操作。如果没有非线性激活函数，无论多少层的Transformer堆叠在一起，整个网络仍然是线性的，这将极大地限制模型的表示能力。
> 

最初的Transformer模型使用的是`ReLU`（Rectified Linear Unit）作为激活函数，但后续的变种和改进中也使用了其他激活函数，如`GELU`（Gaussian Error Linear Unit）。
<img width="424" alt="image" src="https://github.com/pillarliang/blog/assets/28833267/ca78e6bf-ec1d-421e-8e54-c260058b1cb1">

- 第一个线性变换将输入从维度 d 映射到一个更高的维度 4d。
- 第二个线性变换将激活后的输出从维度 4d 映射回原始维度 d。

### Encoder-decoder attention｜编码器-解码器注意力

编码器-解码器注意力是连接编码器跟解码器之间的桥梁。`decoder` 中 `encoder-decoder attention` 的键(key)和值(value)来自 `encoder` 的输出，查询(query)来自 `decoder` 中前一个层的输出。

<img width="702" alt="image" src="https://github.com/pillarliang/blog/assets/28833267/ec1980aa-03d3-4d3d-b40e-b7f78d043c31">

Decoder Block中有2个注意力层的作用: 

- 多头self-attention：用于拟合和利用解码器自身的信息。
- Encoder-Decoder attention：用于整合编码器和解码器之间的信息，以便在解码器中融入编码器的上下文知识并生成准确的输出。

transformer的encoder-decoder框架中,encoder模块会将输入序列转换成一个向量序列(每一个向量对应输入序列中的一个词)。这个向量序列然后会传给decoder。

decoder模块在生成输出序列时,会逐步解码这个来自encoder的向量序列。也就是说,在生成输出序列的每个词时,decoder都会参考encoder输出的整个向量序列,而不是仅仅依赖一个固定的向量。

### **Linear & Softmax**

**Linear：**将解码器输出映射到最终的词典大小维度,作为后续 Softmax 层的输入。

**Softmax：**用来计算输出单词的概率分布。

每一次解码器在产生一个中文字的时候做了一次类似分类的问题。解码器的输出是一个概率的分布，计算标准答案（ground truth）跟分布之间的交叉熵，希望该交叉熵的值越小越好。
<img width="519" alt="image" src="https://github.com/pillarliang/blog/assets/28833267/807f313d-55c9-4122-9376-bf9f9dbebfba">

# seq2seq 模型训练常用技巧

### Teacher Forcing

**what：**

在解码器**训练**的时候，在输入的时候给它正确的答案标准答案给解码器，希望解码器的输出跟正确答案越接近越好。

在测试时不能使用Teacher Forcing，Decoder必须根据先前自己预测的词来预测下一个词。所以训练后期要逐渐减小Teacher Forcing概率,让Decoder习惯自己生成。

<img width="483" alt="image" src="https://github.com/pillarliang/blog/assets/28833267/0af276b1-2980-4c7c-b498-63702eaf5997">

在训练的时候，告诉解码器在已经有 <BOS>、“机”的情况下，要输出“器”，有 <BOS> 、“机”、“器”的情况下输出“学”…

**why：**

- 提高训练速度
- 减少训练过程的误差累积，加快模型收敛

### copy mechanism｜复制机制

对很多任务而言，解码器没有必要自己创造输出，其可以从输入的东西里面复制一些东西。

> 例如：以聊天机器人为例，用户对机器说：“你好，我 是库洛洛”。机器应该回答：“库洛洛你好，很高兴认识你”。机器其实没有必要创造“库洛洛” 这个词汇，
> 

在做摘要的时候，很多的词汇就是直接从原来的文章里面复制出来的，最早有一个从输入复制东西能力的模型叫做指针网络（pointer network），后来还有一个变形叫做复制网络（copy network）。
